lib/doh.c

```diff
@@ -86,12 +86,35 @@ UNITTEST DOHcode doh_encode(const char *host,
   unsigned char *orig = dnsp;
   const char *hostp = host;
 
-  /* The expected output length does not depend on the number of dots within
-   * the host name. It will always be two more than the length of the host
-   * name, one for the size and one trailing null. In case there are dots,
-   * each dot adds one size but removes the need to store the dot, net zero.
+  /* The expected output length is 16 bytes more than the length of
+   * the QNAME-encoding of the host name.
+   *
+   * A valid DNS name may not contain a zero-length label, except at
+   * the end.  For this reason, a name beginning with a dot, or
+   * containing a sequence of two or more consecutive dots, is invalid
+   * and cannot be encoded as a QNAME.
+   *
+   * If the host name ends with a trailing dot, the corresponding
+   * QNAME-encoding is one byte longer than the host name. If (as is
+   * also valid) the hostname is shortened by the omission of the
+   * trailing dot, then its QNAME-encoding will be two bytes longer
+   * than the host name.
+   *
+   * Each [ label, dot ] pair is encoded as [ length, label ],
+   * preserving overall length.  A final [ label ] without a dot is
+   * also encoded as [ length, label ], increasing overall length
+   * by one. The encoding is completed by appending a zero byte,
+   * representing the zero-length root label, again increasing
+   * the overall length by one.
    */
-  const size_t expected_len = 12 + ( 1 + hostlen + 1) + 4;
+  const size_t expected_len =
+    (12                         /* header */
+     + (hostlen ? (hostlen + ((host[hostlen-1] == '.') ? 0 : 1)) : 0)
+     + 1                        /* zero-length root label */
+     + 4);                      /* QTYPE, QCLASS */
```

Comment:
I don't think this needs to deal with a zero hostlen and I think maybe the handling of the trailing dot can be moved outside of the main formula to make this more readable. Maybe like:
~~~c
size_t expected_len;
DEBUGASSERT(hostlen);
expected_len = 12 + 1 + hostlen + 4;
if(host[hostlen-1]!='.')
  hostlen++;

Reply:
If libcurl will never need to query for an RRset at the root, supporting zero hostlen is definitely unnecessary. As I don't know about this, I put it in there, so that this corner case, legitimate from a DNS point of view, would be covered. 

Not doing all the estimation in the initializer certainly improves readability. I think you mean
```
if(host[hostlen-1]!='.')
  expected_len++;
```